PeakNetwork/Core/DecodeInputOperation.swift

```diff
@@ -0,0 +1,242 @@
+    open func decode(input: I) -> Result<O> {
+        fatalError("Subclasses must implement `decode(::)`.")
+    }
+}
+
+
+/// Decode a network response using a `JSONDecoder`.
+open class JSONDecodeOperation<D: Decodable>: DecodeInputOperation<NetworkResponse, D> {
+    
+    public let decoder: JSONDecoder
+
+    /// Create a new `JSONDecodeOperation`.
+    ///
+    /// - Parameters:
+    ///   - input: An optional input `NetworkResponse`.
+    ///   - decoder: The `JSONDecoder` to use when decoding the response data (optional).
+    public init(input: NetworkResponse? = nil,
+                decoder: JSONDecoder = JSONDecoder()) {
+        self.decoder = decoder
+        super.init(input: input)
+    }
+    
+    override open func decode(input: NetworkResponse) -> Result<D> {
+        guard let data = input.data else {
+            return .failure(SerializationError.noData)
+        }
+        
+        return Result {
+            return try decoder.decode(D.self, from: data)
+        }
+    }
+}
+
+/// Decode a network response using a `JSONDecoder`, keeping the `HTTPURLResponse` passed in.
+open class JSONDecodeResponseOperation<D: Decodable>: DecodeInputOperation<NetworkResponse, (D, HTTPURLResponse)> {
+    
+    public let decoder: JSONDecoder
+    
+    /// Create a new `JSONDecodeResponseOperation`.
+    ///
+    /// - Parameters:
+    ///   - input: An optional input `NetworkResponse`.
+    ///   - decoder: The `JSONDecoder` to use when decoding the response data (optional).
+    public init(input: NetworkResponse? = nil,
+                decoder: JSONDecoder = JSONDecoder()) {
+        self.decoder = decoder
+        super.init(input: input)
+    }
+    
+    override open func decode(input: NetworkResponse) -> Result<(D, HTTPURLResponse)> {
+        guard let data = input.data else {
+            return .failure(SerializationError.noData)
+        }
+        
+        return Result {
+            return (try decoder.decode(D.self, from: data), input.urlResponse)
+        }
+    }
+}
+
+/// Decode a network response into a platform-specific Image type
+open class ImageDecodeOperation: DecodeInputOperation<NetworkResponse, PeakImage> {
+    
+    open override func decode(input: NetworkResponse) -> Result<PeakImage> {
+        if let data = input.data, let image = PeakImage(data: data) {
+            return .success(image)
+        } else {
+            return .failure(ImageResponseOperationError.invalidData)
+        }
+    }
+
+    public enum ImageResponseOperationError: Error {
+        /// Could not initialize the image from the specified data.
+        case invalidData
+    }
+}
+
+
+
+/// Perform two operations, passing the result of the first to the second.
+/// The input to this operation is used as the input to the first operation.
+/// The output of the second operation is used as the output of this operation.
+open class SequenceOperation<First, Second>: ConcurrentOperation, ProducesResult
```

Comment:
This seems like a more specialised version of `GroupOperation`. I understand this provides output from the second operation, but if possible, I think it would be nicer to modify `GroupOperation` to provide output from the final operation and use it in place of `SequenceOperation`.

Reply:
This would be nice but I can't figure out a way to allow GroupOperation to take input as well as produce output, without having initialisers with set numbers of arguments. I can't just take an array of operations and `passesResult:` between them all, basically.

Reply:
Okay. This is only required for the now deprecated `DecodableOperations`, right? If so, then it's not really an issue.